File: C++/src/group_anagrams.cpp
#include <algorithm>
#include <group_anagrams.hpp>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

vector<vector<string>> Solution::groupAnagrams(vector<string> &strs) {
  unordered_map<string, vector<string>> hash_map;

  for (string s : strs) {
    string word = s;
    sort(word.begin(), word.end());
    hash_map[word].push_back(s);
  }

  vector<vector<string>> result;
  for (auto pair : hash_map) {
    result.push_back(pair.second);
  }

  return result;
}

File: JavaScript/src/group_anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
function groupAnagrams(strs) {
  /**@type {Map<string, string[]>} */
  const map = new Map();

  for (const s of strs) {
    const key = s.split("").sort().join("");
    const group = map.get(key);
    if (group !== undefined) {
      group.push(s);
    } else {
      const newGroup = [s];
      map.set(key, newGroup);
    }
  }

  /** @type {string[][]} */
  const result = [];
  map.forEach((v) => {
    result.push(v);
  });

  return result;
}

export { groupAnagrams };
